arch/arm/core/aarch64/reset.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2021 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_internal.h>
+#include "vector_table.h"
+
+void __weak z_arm64_el_highest_plat_init(void)
+{
+	/* do nothing */
+}
+
+void __weak z_arm64_el3_plat_init(void)
+{
+	/* do nothing */
+}
+
+void __weak z_arm64_el2_plat_init(void)
+{
+	/* do nothing */
+}
+
+void __weak z_arm64_el1_plat_init(void)
+{
+	/* do nothing */
+}
+
+void z_arm64_el_highest_init(void)
+{
+	write_cntfrq_el0(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+
+	z_arm64_el_highest_plat_init();
+
+	isb();
+}
+
+void z_arm64_el3_init(void)
+{
+	uint64_t reg;
+
+	/* Setup vector table */
+	write_vbar_el3((uint64_t)_vector_table);
+	isb();
+
+	reg = read_cptr_el3();
+	reg &= ~(CPTR_EL3_RES0 |	/* RES0 */
+		 CPTR_EL3_TTA_BIT |	/* Do not trap sysreg accesses */
+		 CPTR_EL3_TFP_BIT |	/* Do not trap SVE, SIMD and FP */
+		 CPTR_EL3_TCPAC_BIT);	/* Do not trap CPTR_EL2 / CPACR_EL1 accesses */
+	reg |= CPTR_EL3_EZ_BIT;		/* Do not trap SVE accesses */
+	write_cptr_el3(reg);
+
+	reg = read_scr_el3();
+	reg &= ~(SCR_NS_BIT |		/* XXX: Boot in secure mode */
+		 SCR_IRQ_BIT |		/* Do not route IRQs to EL3 */
+		 SCR_FIQ_BIT |		/* Do not route FIQs to EL3 */
+		 SCR_EA_BIT);		/* Do not route Serrors to EL3 */
+	reg |= (SCR_RES1 |		/* RES1 */
+		SCR_RW_BIT |		/* EL2 execution state is AArch64 */
+		SCR_ST_BIT |		/* Do not trap EL1 accesses to timer */
+		SCR_HCE_BIT |		/* Do not trap HVC */
+		SCR_SMD_BIT);		/* Do not trap SMC */
+
+	write_scr_el3(reg);
+
+	z_arm64_el3_plat_init();
+
+	isb();
+}
+
+void z_arm64_el2_init(void)
+{
+	uint64_t reg;
+
+	reg = read_sctlr_el2();
+	reg |= (SCTLR_EL2_RES1 |	/* RES1 */
+		SCTLR_I_BIT |		/* Enable i-cache */
+		SCTLR_SA_BIT);		/* Enable SP alignment check */
+	write_sctlr_el2(reg);
+
+	reg = read_hcr_el2();
+	reg |= HCR_RW_BIT;		/* EL1 Execution state is AArch64 */
+	write_hcr_el2(reg);
+
+	zero_cntvoff_el2();		/* Set 64-bit virtual timer offset to 0 */
+	zero_cnthctl_el2();
+	zero_cnthp_ctl_el2();
+
+	z_arm64_el2_plat_init();
+
+	isb();
+}
+
+void z_arm64_el1_init(void)
+{
+	uint64_t reg;
+
+	/* Setup vector table */
+	write_vbar_el1((uint64_t)_vector_table);
+	isb();
+
+	write_cpacr_el1(CPACR_EL1_FPEN_NOTRAP); /* Do not trap NEON/SIMD/FP */
+
+	reg = read_sctlr_el1();
+	reg |= (SCTLR_EL1_RES1 |		/* RES1 */
+		SCTLR_I_BIT |			/* Enable i-cache */
+		SCTLR_SA_BIT);			/* Enable SP alignment check */
+	write_sctlr_el1(reg);
+
+	z_arm64_el1_plat_init();
+
+	isb();
+}
+
+void z_arm64_el3_get_next_el(uint64_t switch_addr)
+{
+	uint64_t spsr;
+
+	write_elr_el3(switch_addr);
+
+	/* Mask the DAIF */
+	spsr = SPSR_DAIF_MASK;
+
+	/*
+	 * Is considered an illegal return "[..] a return to EL2 when EL3 is
+	 * implemented and the value of the SCR_EL3.NS bit is 0 if
+	 * ARMv8.4-SecEL2is not implemented" (D1.11.2 from ARM DDI 0487E.a)
+	 *
+	 * XXX: boot with SCR_EL3.NS == 0
+	 */
+	if (is_el_implemented(2) && is_el2_sec_supported()) {
```

Comment:
This... looks wrong to me.

> `is_el2_sec_supported`

EL2 is _never_ in secure mode. I'll have to take a look as to what `is_el2_sec_supported` actually does...

Reply:
`EL2` can be is secure mode if the CPU supports `ARMv8.4-SecEL2`. In that case dropping from `EL3` to `EL2` with `SCR_EL3.NS == 0` is considered illegal. Since we are currently booting with `SRC_EL3.NS==0` I have to check this.

In fact also QEMU refuses to drop from `EL3` to `EL2` when booted with `secure=on` and `SCR_EL3.NS==0`, it simply stays in `EL3`.

Reply:
> In that case dropping from EL3 to EL2 with SCR_EL3.NS == 0 is considered illegal.

This is actually the case in the base spec too:

> The ARMv8 architecture has a generic mechanism for handling returns to a mode or state that is illegal. In AArch64
state, this can occur as the result of any of the following situations:
> [...]
> A return to EL2 when EL3 is implemented and the value of the SCR_EL3.NS bit is 0

I am however rather confused.

If you can't drop from EL3 to EL2 with SCR_EL3.NS == 0 even with `ARMv8.4-SecEL2` support, then how does `ARMv8.4-SecEL2` actually get to secure-mode EL2?

(Note negative - NS=0 means not not secure means secure.)

Reply:
> If you can't drop from EL3 to EL2 with SCR_EL3.NS == 0 even with `ARMv8.4-SecEL2` support, then how does `ARMv8.4-SecEL2` actually get to secure-mode EL2?

Where do you infer this from? Reading the spec it looks to me that you can drop from `EL3` to `EL2` if:
1. You are in secure mode (that is `SCR_EL3.NS == 0`)
2. You do support `ARMv8.4-SecEL2`

In our case we are missing the point 2 but we have the point 1 (at least for now), so we cannot drop from `EL3` to `EL2`.

Reply:
Oooooh, my bad. I hope we are on the same page now :)